### DISCUSSION TITLE
Output valid TileJSON for PostgreSQL Function Sources

### DISCUSSION THREAD

**Discussion by ohrie** [maplibre/martin#1214] [Q&A]
When using [the (simple function) example](https://maplibre.org/martin/sources-pg-functions.html) from martin docs for PostgreSQL Function Sources, the resulting TileJSON looks like this:

```json
{
  "tilejson": "3.0.0",
  "tiles": [
    "https://<YOURTILESERVER>.com/function_zxy_query/{z}/{x}/{y}"
  ],
  "description": "public.function_zxy_query",
  "name": "function_zxy_query"
}
```

This is not a valid TileJSON, because the [TileJSON spec](https://github.com/mapbox/tilejson-spec/tree/master/3.0.0#33-vector_layers) requires the `fields` attribute to be present. Therefore the usage in MapLibre fails.

I currently don't know how to implement an according solution. The solution modyfing the TileJSON via the `COMMENT` SQL statement further down the docs didn't work.

**Comment by sharkAndshark** [comment]
Hi @ohrie 
I tried this on the test database and it still works. Would you mind share you SQL here?
```sql
COMMENT ON FUNCTION "public"."function_zxy_query" ( "z" int4, "x" int4, "y" int4, "query" json ) IS '
{
	"description": null,
	"fields":{
		"Gid":"int4",
		"TABLE":"text"
	},
	"foo": {"bar": "foo"}
}
';
```

```shell
async@LAPTOP-PLI51BMO:~/repos/martin$ curl http://localhost:3000/function_zxy_query
{"tilejson":"3.0.0","tiles":["http://localhost:3000/function_zxy_query/{z}/{x}/{y}"],"name":"function_zxy_query","fields":{"Gid":"int4","TABLE":"text"},"foo":{"bar":"foo"}}
```

**Comment by nyurik** [comment]
@ohrie you are correct that, per specification, tilejson must include `vector_layers` value that has a `fields` sub-value. The [example](https://maplibre.org/martin/sources-pg-functions.html#tilejson-in-sql-comments) in the documentation actually does that. There is one issue:  in the [Modifying TileJSON](https://maplibre.org/martin/sources-pg-functions.html#modifying-tilejson) section it says that it generates tilejson automatically, but does not say that it is not a fully valid tilejson unless the content of the tile is non-vector (but at the moment there is no way for the user to specify that).

We cannot really fix this automatically because Martin has no idea which tiles are available. Only after the first tiles have been requested, we could, in theory, wait until the first tiles start being generated, parse them, figure out available fields, and update `vector_layers`, but this feels like a very hacky, imperfect, and unstable solution (tiles may not include all possible fields).
